### PR TITLE
XML String Precision

### PR DESCRIPTION
**Pull request by fcanderson**
While there are various constructors available in class ```SimTK::String``` that allowed for specification of output format, the templatized constructor for class ```String``` (i.e., ```String::String(const T& t)```) was fixed at the output precision set by ```std::ostream```. As a consequence, any ```String``` conversion for a type ```T``` for which there exists no matching constructor was fixed to an output precision of ```6```.

So that output precision (i.e., the number of significant figures used to represent a value) can be specified for such types, this commit adds an optional ```precision``` argument to the templatized ```String``` constructor and to the methods in class ```SimTK::Xml``` that call this templatized ```String``` constructor. In particular, the following methods now include the optional ```precision``` argument:
```
SimTK::String::String<T>(const T& t, int precision = String::DefaultOutputPrecision)

SimTK::Xml::Element::Element<T>(const String& tagWord, const T& value, int precision = String::DefaultOutputPrecision)

SimTK::Xml::Element::setValueAs<T>(const T& value, int precision = String::DefaultOutputPrecision)
```
Because the ```precision``` argument is optional, no API adjustments are needed in existing code that calls the above methods.

The static constant ```String::DefaultOutputPrecision``` was added to support this new functionality. Its value is chosen to be ```6``` to be consistent with the default precision of class ```std::ostream``` and so that previous behavior would be maintained.

The primary benefit of this commit is that a broad range of numerical value types in SimTK (e.g., ```Vec3```, ```Vec4```, etc.) can be serialized via class ```SimTK::Xml``` with a specified number of significant figures.  The changes revert default output precision to what it was prior to #762 but also make it possible to serialize ```SimTK::State```, among other quantities, in lossless fashion.

User facing Doxygen comments were updated and a subunit test (```testOutputPrecision```) was added to ```testXml.cpp```.

- - -
This change is [Reviewable](https://reviewable.io/reviews/simbody/simbody/776)

### CODE REVIEW

**Review by sherm1 — APPROVED** [review]
Thanks, Clay. :lgtm: pending a few minor fixes

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/776#-:-NqPtZ-E3WhzhxIknC3m:b-rl04zb)* status: all files reviewed, 7 unresolved discussions (waiting on @fcanderson)

___
*[`SimTKcommon/include/SimTKcommon/internal/String.h` line 177 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPoScV4VTSOrvmSH9l:-NqPoScV4VTSOrvmSH9m:b-d5ul3n) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/include/SimTKcommon/internal/String.h#L177)):*
> ```objc
> //     around a default output precision of 6 (e.g., model files, GUI
> //     displays).
> //   * The default output precision should be great than or equal to 1 and
> ```

typo: great -> greater
___
*[`SimTKcommon/include/SimTKcommon/internal/String.h` line 181 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPpzYw2EcLt9EDh3lG:-NqPpzYw2EcLt9EDh3lH:b-lk0u6c) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/include/SimTKcommon/internal/String.h#L181)):*
> ```objc
> // - Because this constant is used only to supply the default value of an
> //   optional 'precision' argument in the following templatized methods,
> //   no API adjustments are required in code that calls these methods. 
> ```

minor: the concept of "API adjustment" here only applies to a single release -- it won't mean anything later. This makes sense in the commit message but not so much here. Sufficient to say that the intent is to match the default output precision of 6 as you did above.
___
*[`SimTKcommon/tests/TestXml.cpp` line 345 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPs1IN4ZK6Hr6onFVR:-NqPs1IN4ZK6Hr6onFVS:b-f7ywrt) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L345)):*
> ```cpp
>     // will result in a consistent value that utilizes all available digits
>     // whether the type of 'input' is float or double.
>     // A Vec<1> is used so that the templatized String constructor is evoked.
> ```

nit: evoked -> invoked
___
*[`SimTKcommon/tests/TestXml.cpp` line 348 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPsAMPAbAzjiaPinhf:-NqPsAMPAbAzjiaPinhg:b-4d08kt) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L348)):*
> ```cpp
>     SimTK::Vec<1> input(0.123456789012345678901234);
> 
>     // Expected conversions to String for a range of precsions -----
> ```

typo: precsions -> precisions
___
*[`SimTKcommon/tests/TestXml.cpp` line 354 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPsb8b16YBrANdCi4H:-NqPsb8b16YBrANdCi4I:bmb0i83) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L354)):*
> ```cpp
>     // Though not tested past p = 7, the following rounding behavior is what
>     // is observed for a double.
>     const int n{24};
> ```

BTW consider naming this something more suggestive like `num_precisions`
___
*[`SimTKcommon/tests/TestXml.cpp` line 432 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPsvZfCDsG1Id7SKzY:-NqPsvZfCDsG1Id7SKzZ:b-ej5sof) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L432)):*
> ```cpp
>     //std::ostringstream os;
>     //String outputDefault =
>     //    stringStreamInsertHelper<Vec<1>>(os, input, true).str();
> ```

minor: remove dead code before merging
___
*[`SimTKcommon/tests/TestXml.cpp` line 434 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPszWCCEJvuwmhVLWF:-NqPszWCCEJvuwmhVLWG:b-pl60ox) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L434)):*
> ```cpp
>     //    stringStreamInsertHelper<Vec<1>>(os, input, true).str();
> 
>     // Test when pecision is greater than LosslessNumDigits.
> ```

typo: pecision

**Review by fcanderson — COMMENTED** [review]
Thanks for the review, Sherm. Lots of typo catches. I believe I took all your suggestions. You might want to take a second look at the developer notes on ```String::DefaultOutputPrecision``` (String.h, lines 168-183) though.

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/776#-:-NqT0UkYDXNSMbgp6NGa:b-s3neik)* status: 1 of 3 files reviewed, 7 unresolved discussions (waiting on @sherm1)

___
*[`SimTKcommon/include/SimTKcommon/internal/String.h` line 177 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPoScV4VTSOrvmSH9l:-NqSmcV539A6ry0OD8A9:b-r0jbje) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/include/SimTKcommon/internal/String.h#L177)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

typo: great -> greater
</blockquote></details>

corrected. done.
___
*[`SimTKcommon/include/SimTKcommon/internal/String.h` line 181 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPpzYw2EcLt9EDh3lG:-NqSrwSG9UpYJ_Na-w6r:b-x6s7x1) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/include/SimTKcommon/internal/String.h#L181)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

minor: the concept of "API adjustment" here only applies to a single release -- it won't mean anything later. This makes sense in the commit message but not so much here. Sufficient to say that the intent is to match the default output precision of 6 as you did above.
</blockquote></details>

Yep. I see what you mean (I think) and removed the part about "API Adjustments". I do think there is some value in listing the methods in which ```String::DefaultOutputPrecision``` is used, but maybe that just opens the door for the notes to become out of date due to changes introduced down the road. I took a shot at simplifying the notes. See if the revision looks better to you or if perhaps the notes are still too tied to the history of the changes.
___
*[`SimTKcommon/tests/TestXml.cpp` line 345 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPs1IN4ZK6Hr6onFVR:-NqSwWL49CMP0u6IGFsH:b-r0jbje) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L345)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

nit: evoked -> invoked
</blockquote></details>

corrected. done.
___
*[`SimTKcommon/tests/TestXml.cpp` line 348 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPsAMPAbAzjiaPinhf:-NqSy7qNBhmNJPdsgeqW:b-r0jbje) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L348)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

typo: precsions -> precisions
</blockquote></details>

corrected. done.
___
*[`SimTKcommon/tests/TestXml.cpp` line 354 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPsb8b16YBrANdCi4H:-NqSzNTC2TMKH7Jwev01:b4bi9iu) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L354)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider naming this something more suggestive like `num_precisions`
</blockquote></details>

I took your suggestion. done.
___
*[`SimTKcommon/tests/TestXml.cpp` line 432 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPsvZfCDsG1Id7SKzY:-NqT-69u5Q3213oSV73m:b-m9js0x) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L432)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

minor: remove dead code before merging
</blockquote></details>

removed. done.
___
*[`SimTKcommon/tests/TestXml.cpp` line 434 at r1](https://reviewable.io/reviews/simbody/simbody/776#-NqPszWCCEJvuwmhVLWF:-NqT0IIEB1kh_bwLESaf:b-r0jbje) ([raw file](https://github.com/simbody/simbody/blob/a48e986341db65a06115db33e6449c1272a786cc/SimTKcommon/tests/TestXml.cpp#L434)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

typo: pecision
</blockquote></details>

corrected. done.

**Review by fcanderson — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/776#-:-NqTBptP4wDc5WYC66E8:bnfp4nl)* status: 1 of 3 files reviewed, 7 unresolved discussions (waiting on @sherm1)

**Review by sherm1 — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/776)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @fcanderson)

**Pull request merged**
